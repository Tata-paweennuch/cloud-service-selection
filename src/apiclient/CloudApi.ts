import { AxiosResponse } from 'axios'
import axios from 'axios';

export interface Cloud {
  cloud_description: string,
  cloud_name: string,
  geo_latitude: number,
  geo_longitude: number,
  geo_region: string,
  distance?: number  // added when calculating distance between 2 plaes
}

export class CloudApi {
  // making this to be a method because .....
  private getBaseUrl(): string {
    return 'https://api.aiven.io/v1/clouds'
  }

  public async getCloudList(): Promise<Cloud[]> {
    const url = this.getBaseUrl()

    const response: AxiosResponse = await axios.get(url)
    return response?.data.clouds || []
  }
}
