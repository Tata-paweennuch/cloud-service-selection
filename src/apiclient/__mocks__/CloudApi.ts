export class CloudApi {
  public async getCloudList() {
    return [
      {
        cloud_description: "Europe, Finland - Google Cloud: Finland",
        cloud_name: "google-europe-north1",
        geo_latitude: 60.5693,
        geo_longitude: 27.1878,
        geo_region: "europe"
      },
      {
        cloud_description: "Asia, Japan - Azure: Japan East",
        cloud_name: "azure-japaneast",
        geo_latitude: 35.68,
        geo_longitude: 139.68,
        geo_region: "east asia"
      },
      {
        cloud_description: "United States, California - Amazon Web Services: N. California",
        cloud_name: "aws-us-west-1",
        geo_latitude: 37.78,
        geo_longitude: -122.42,
        geo_region: "north america"
      }
    ]
  }
}
