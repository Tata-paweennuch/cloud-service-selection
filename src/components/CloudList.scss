.content-wrapper {
  padding: 20px 40px;
  background-color: rgba(15,17,27,0.4);
  border-radius: 14px;
  width: 620px;
}

.dropdown {
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-bottom: 1.5rem;

  label {
    display: inline-block;
    width: 100%;
    padding: 10px 0;
  }

  select {
    width: 100%;
    outline: none;
    background: none;
    border: none;
    padding: 10px;
    cursor: pointer;

    border: 1px solid rgb(51, 48, 48);
    border-radius: 14px;
    background-color: rgb(196,227,206);
    opacity: 0.6;
  }
}
