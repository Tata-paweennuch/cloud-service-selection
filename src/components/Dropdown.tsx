import { Component } from 'react';
import { Option } from './SharedInterface'

interface DropdownProps {
  label: string,
  id: string,
  value: string,
  handleOnChange: (name: string) => void,
  options: Option[]
}

class Dropdown extends Component<DropdownProps, {}> {
  handleValueChange = (e: any) => {
    this.props.handleOnChange(e.target.value)
  }

  render() {
    const { label, id, value, options } = this.props

    // try to use semantic elements, so we can get the best accessibility
    return (
      <div className="dropdown">
        <label htmlFor={id}>{ label }</label>
        <select name={id} id={id} onChange={this.handleValueChange} value={value}>
          {options.map((option, index) =>
            <option value={option.value} key={option.value + index}>{option.displayName}</option>
          )}
        </select>
      </div>  
    )  
  }
}

export default Dropdown;
