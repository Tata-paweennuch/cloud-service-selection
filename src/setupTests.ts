// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom'

import { configure } from 'enzyme'
// https://github.com/enzymejs/enzyme/issues/2429
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'

configure({ adapter: new Adapter() })

import axios from 'axios'
jest.mock("axios")

export const mockedAxios = axios as jest.Mocked<typeof axios>

// https://stackoverflow.com/a/51829561
export const getCurrentPositionMock = jest.fn()
  .mockImplementationOnce((success) => Promise.resolve(success({
    coords: {
      latitude: 51.1,
      longitude: 45.3
    }
  })))
const mockGeolocation = {
  getCurrentPosition: getCurrentPositionMock
}
global.navigator.geolocation = mockGeolocation
