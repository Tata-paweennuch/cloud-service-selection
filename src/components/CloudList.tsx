import React, { Component } from 'react';
import './CloudList.scss';
import { CloudApi, Cloud } from '../apiclient/CloudApi'
import { getDistanceFromLatLonInKm } from '../helper/DistanceCalculation'
import Dropdown from './Dropdown'
import { Option } from './SharedInterface'

interface ICloudListState {
  clouds: Cloud[],
  isLoading: boolean,
  cloudNameOptions: Option[],
  regionOptions: Option[],
  selectedCloud: string,
  selectedRegion: string,
  selectedCity: string,
  filteredItems: Cloud[],
  location: {
    lat: number,
    lon: number
  }
}

export const emptyOption: Option = { value: '', displayName: '-- select an option --' }

class CloudList extends Component<{}, ICloudListState> {

  state: ICloudListState = {
    clouds: [],
    isLoading: false,
    cloudNameOptions: [emptyOption],
    regionOptions: [emptyOption],
    selectedCloud: '',
    selectedRegion: '',
    selectedCity: '',
    filteredItems: [],
    location: {
      // default to Helsinki's location
      lat: 60.1699,
      lon: 24.9384
    }
  }

  componentDidMount() {
    const cloudApi = new CloudApi()
    this.setState({ isLoading: true }, () => {
      cloudApi.getCloudList()
        .then(res => {
          this.setState({ clouds: res || [], isLoading: false })
          this.setCloudOptions()
          this.setRegionOptions()
          // now selectedCloud and selectedRegion are empty, so show all cloud options
          this.filterCloudByNameAndOrRegion()
        }).catch(() => {
          this.setState({ clouds: [], isLoading: false })
        })
    })

    // get user's location by using the HTML5 geolocation API
    // ref: https://stackoverflow.com/questions/13840516/how-to-find-my-distance-to-a-known-location-in-javascript
    // improvement: show pop-up alert to ask for consent
    window.navigator.geolocation.getCurrentPosition((pos) => {
      this.setState({ location: { lat: pos.coords.latitude, lon: pos.coords.longitude } })
    })
  }

  filterCloudsByName = (name: string): Cloud[] => {
    const filtered = this.state.clouds.filter((cloud: Cloud) => {
      return cloud.cloud_name.includes(name)
    })
    return filtered
  }

  filterCloudByRegion = (region: string, list: Cloud[]): Cloud[] => {
    const filtered = list.filter((item) => item.geo_region === region)
    return filtered
  }

  filterCloudByNameAndOrRegion = (): void => {
    const name = this.state.selectedCloud
    const region = this.state.selectedRegion

    // if name only --> filter name 
    // if region only --> filter region
    // if name & region -->  filter name then region
    // if no name & region --> return the whole cloud list from API
    let filtered = this.state.clouds
    if(name) {
      filtered = this.filterCloudsByName(name)
    }
    if(region) {
      filtered = this.filterCloudByRegion(region, filtered)
    }

    this.setState({ filteredItems: filtered }, () => {
      this.sortDistanceAscen()
    })
  }

  getCloudFullName = (name: string):string => {
    switch (name) {
      case 'aws':
        return 'Amazon Web Services'
      case 'azure':
        return 'Microsoft Azure'
      case 'google':
        return 'Google Cloud'
      case 'do':
        return 'DigitalOcean'
      case 'upcloud':
        return 'UpCloud'
      default:
        return name
    }
  }

  setCloudOptions = (): void => {
    if(this.state.clouds.length > 0) {
      const names = this.state.clouds.map((cloud: Cloud) => cloud.cloud_name.split('-')[0])
      const uniqueNames = Array.from(new Set(names))
      const formattedOptions = [] as Option[]
      uniqueNames.forEach((name) => {
        formattedOptions.push({ value: name, displayName: this.getCloudFullName(name)  })
      })
      this.setState({
        cloudNameOptions: [emptyOption, ...formattedOptions]
      })
    } else {
      this.setState({ cloudNameOptions: [emptyOption]})
    }
  }

  setRegionOptions = (): void => {
    if(this.state.clouds.length > 0) {
      const regions = this.state.clouds.map(cloud => cloud.geo_region)
      const uniqueRegions = Array.from(new Set(regions))
      const formattedOptions = uniqueRegions.map(region => {
        const displayName = region.toLowerCase()
          .split(' ')
          .map((string) => string.charAt(0).toUpperCase() + string.substring(1))
          .join(' ')
        return { value: region, displayName }
      })
      this.setState({ regionOptions: [emptyOption, ...formattedOptions] })
    } else {
      this.setState({ regionOptions: [emptyOption] })
    }
  }

  handleCloudNameChange = (cloudName: string): void => {
    this.setState({ selectedCloud: cloudName }, () => {
      this.filterCloudByNameAndOrRegion()
    })
  }

  handleRegionChange = (region: string): void => {
    this.setState({ selectedRegion: region }, () => {
      this.filterCloudByNameAndOrRegion()
    })
  }

  handleCityChange = (e: React.SyntheticEvent): void => {
    const target  = e.target as HTMLSelectElement
    this.setState({ selectedCity: target.value })
  }

  calculateDistances = (): Cloud[] => {
    let filteredList = this.state.filteredItems
    if(filteredList.length > 1) {
      filteredList.forEach((item: Cloud, index) => {
        filteredList[index].distance = getDistanceFromLatLonInKm(item.geo_latitude, item.geo_longitude, this.state.location.lat, this.state.location.lon)
      })
    }
    // otherwise, no need to sort and cal distances
    
    return filteredList  
  }

  sortDistanceAscen = (): void => {
    let data = this.calculateDistances()
    
    let sortedData = data.sort((a, b) => (a.distance? a.distance : 0) - (b.distance? b.distance : 0))
    this.setState({ filteredItems: sortedData })
  }

  render() {    
    const filteredResult = () => {
      if(this.state.filteredItems.length > 0) {
        return this.state.filteredItems.map((cloud: Cloud, index: number) =>
          <option value={cloud.cloud_description} key={cloud.cloud_description + index}>{cloud.cloud_description}{(this.state.filteredItems.length !== 1 && index === 0)? ' * Nearest' : ''}</option>
        )
      } else {
        return <option value=''>No cloud service available with your criteria</option>
      }
    }
    
    const selection = (
      <React.Fragment>
        <Dropdown 
          label="Select cloud"
          id="cloud"
          value={this.state.selectedCloud}
          handleOnChange={this.handleCloudNameChange}
          options={this.state.cloudNameOptions} />
        <Dropdown 
          label="Select region"
          id="region"
          value={this.state.selectedRegion}
          handleOnChange={this.handleRegionChange}
          options={this.state.regionOptions} />
        <div className="dropdown">
          <label htmlFor="city">
            Select city
            <small>(Listed from nearest to farest)</small>
          </label>
          <select name="city" id="city" onChange={this.handleCityChange} value={this.state.selectedCity}>
            {filteredResult()}
          </select>
        </div>
      </React.Fragment>
    )
    
    return(
      <div className="content-wrapper">
        {this.state.isLoading? <p>Loading ....</p> : selection}
      </div>
    );
  }
}

export default CloudList;
