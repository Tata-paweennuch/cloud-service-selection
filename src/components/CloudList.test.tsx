import { shallow } from 'enzyme';
import CloudList, { emptyOption } from './CloudList';
import { CloudApi, Cloud } from '../apiclient/CloudApi'
import * as helper from '../helper/DistanceCalculation'
import Dropdown from './Dropdown'
import { getCurrentPositionMock } from '../setupTests'
import renderer from 'react-test-renderer'

let wrapper: any

const mockClouds = [
  {
    cloud_description: "Europe, Finland - Google Cloud: Finland",
    cloud_name: "google-europe-north1",
    geo_latitude: 60.5693,
    geo_longitude: 27.1878,
    geo_region: "europe"
  },
  {
    cloud_description: "Asia, Japan - Azure: Japan East",
    cloud_name: "azure-japaneast",
    geo_latitude: 35.68,
    geo_longitude: 139.68,
    geo_region: "east asia"
  },
  {
    cloud_description: "United States, California - Amazon Web Services: N. California",
    cloud_name: "aws-us-west-1",
    geo_latitude: 37.78,
    geo_longitude: -122.42,
    geo_region: "north america"
  }
]

// has a same cloud name in 2 object(google)
const mockClouds2 = [
  ...mockClouds,
  {
    cloud_description: "Canada, Quebec - Google Cloud: MontrÃ©al",
    cloud_name: "google-northamerica-northeast1",
    geo_latitude: 45.5,
    geo_longitude: -73.57,
    geo_region: "north america"
  }
]

// use the mock class instead of the real one
jest.mock('../apiclient/CloudApi');

afterEach(() => {
  if(wrapper) {
    wrapper.unmount()
    wrapper = null
  }
})

test('should render correctly when isLoading is true', async () => {
  const component = shallow<CloudList>(<CloudList />)
  await component.instance().componentDidMount()
  component.setState({
    isLoading: true
  })

  const tree = renderer.create(component.getElement()).toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render correctly with cloud selection', async () => {
  const tree = renderer.create(<CloudList />)
  await CloudApi.prototype.getCloudList()
  expect(tree.toJSON()).toMatchSnapshot()
})

test('should have correct default state properties', () => {
  const wrapper = shallow(<CloudList />, { disableLifecycleMethods: true })
  expect(wrapper.state('isLoading')).toBeFalsy()
  expect(wrapper.state('clouds')).toEqual([])
  expect(wrapper.state('cloudNameOptions')).toEqual([emptyOption])
  expect(wrapper.state('regionOptions')).toEqual([emptyOption])
  expect(wrapper.state('selectedCloud')).toEqual('')
  expect(wrapper.state('selectedRegion')).toEqual('')
  expect(wrapper.state('selectedCity')).toEqual('')
  expect(wrapper.state('filteredItems')).toEqual([])
  expect(wrapper.state('location')).toEqual({ lat: 60.1699, lon: 24.9384 })
})

test('should fetch cloud data during componentDidMount hook', async () => {
  wrapper = shallow(<CloudList />)
  const apiCallSpy = jest.spyOn(CloudApi.prototype, 'getCloudList')
  wrapper.instance().setCloudOptions = jest.fn()
  wrapper.instance().setRegionOptions = jest.fn()
  wrapper.instance().filterCloudByNameAndOrRegion = jest.fn()
  wrapper.instance().forceUpdate()
  await wrapper.instance().componentDidMount()

  expect(CloudApi.prototype.getCloudList).toHaveBeenCalled()
  expect(apiCallSpy).toHaveBeenCalled()
  expect(wrapper.instance().setCloudOptions).toHaveBeenCalled()
  expect(wrapper.instance().setRegionOptions).toHaveBeenCalled()
  expect(wrapper.instance().filterCloudByNameAndOrRegion).toHaveBeenCalled()
  expect(wrapper.state('clouds')).toEqual(mockClouds)
  expect(getCurrentPositionMock).toHaveBeenCalled()

  apiCallSpy.mockReset()
  apiCallSpy.mockRestore()
})

test('should show loading text when isLoading state is true', () => {
  wrapper = shallow(<CloudList />)
  wrapper.setState({ isLoading: true })

  expect(wrapper.find('.content-wrapper').text()).toContain('Loading')
})

test('should render Child components', async() => {
  wrapper = shallow(<CloudList />)
  await wrapper.instance().componentDidMount()
  expect(wrapper.find(Dropdown).length).toEqual(2)

  const cloudSelection = wrapper.find(Dropdown).at(0)
  const expectedCloudOptions = [
    { value: 'google', displayName: 'Google Cloud' },
    { value: 'azure', displayName: 'Microsoft Azure' },
    { value: 'aws', displayName: 'Amazon Web Services' }
  ]
  expect(cloudSelection.props()['label']).toEqual('Select cloud')
  expect(cloudSelection.props()['id']).toEqual('cloud')
  expect(cloudSelection.props()['value']).toEqual('')
  expect(cloudSelection.props()['handleOnChange']).toEqual(wrapper.instance().handleCloudNameChange)
  expect(cloudSelection.props()['options']).toEqual([emptyOption, ...expectedCloudOptions])

  const regionSelection = wrapper.find(Dropdown).at(1)
  // since region options based on select cloud name which we set default to mockClouds[0].name(google)
  const expectedRegionOptions = [
    { value: mockClouds[0].geo_region, displayName: 'Europe' },
    { value: mockClouds[1].geo_region, displayName: 'East Asia' },
    { value: mockClouds[2].geo_region, displayName: 'North America' }
  ]
  expect(regionSelection.props()['label']).toEqual('Select region')
  expect(regionSelection.props()['id']).toEqual('region')
  expect(regionSelection.props()['value']).toEqual('')
  expect(regionSelection.props()['handleOnChange']).toEqual(wrapper.instance().handleRegionChange)
  expect(regionSelection.props()['options']).toEqual([emptyOption, ...expectedRegionOptions])
})

test('should display the city selector correctly', async () => {
  wrapper = shallow(<CloudList />)
  await wrapper.instance().componentDidMount()
  wrapper.setState({ filteredItems: mockClouds })

  expect(wrapper.find('.dropdown label').text()).toEqual('Select city(Listed from nearest to farest)')
  expect(wrapper.find('.dropdown label').props()['htmlFor']).toEqual('city')
  expect(wrapper.find('.dropdown select').props()['name']).toEqual('city')
  expect(wrapper.find('.dropdown select').props()['id']).toEqual('city')
  expect(wrapper.find('.dropdown select').props()['onChange']).toEqual(wrapper.instance().handleCityChange)
  expect(wrapper.find('.dropdown select').props()['value']).toEqual(wrapper.state('selectedCity'))
  // has 3 options
  expect(wrapper.find('.dropdown select option').length).toEqual(3)
  expect(wrapper.find('.dropdown select option').at(0).text()).toEqual(mockClouds[0].cloud_description + ' * Nearest')
  expect(wrapper.find('.dropdown select option').at(0).props()['value']).toEqual(mockClouds[0].cloud_description)
  expect(wrapper.find('.dropdown select option').at(1).text()).toEqual(mockClouds[1].cloud_description)
  expect(wrapper.find('.dropdown select option').at(1).props()['value']).toEqual(mockClouds[1].cloud_description)
  expect(wrapper.find('.dropdown select option').at(2).text()).toEqual(mockClouds[2].cloud_description)
  expect(wrapper.find('.dropdown select option').at(2).props()['value']).toEqual(mockClouds[2].cloud_description)
})

describe('Filter methods', () => {
  test('should filter cloud list by cloud name', () => {
    wrapper = shallow(<CloudList />)
    wrapper.setState({
      clouds: mockClouds
    })
    const expectedResult = [{
      cloud_description: "United States, California - Amazon Web Services: N. California",
      cloud_name: "aws-us-west-1",
      geo_latitude: 37.78,
      geo_longitude: -122.42,
      geo_region: "north america"
    }]
  
    expect(wrapper.instance().filterCloudsByName('aws')).toEqual(expectedResult)
    expect(wrapper.instance().filterCloudsByName('something')).toEqual([])
  })
  
  test('should filter cloud list by region', () => {
    wrapper = shallow(<CloudList />)
    const expectedResult = [{
      cloud_description: "Asia, Japan - Azure: Japan East",
      cloud_name: "azure-japaneast",
      geo_latitude: 35.68,
      geo_longitude: 139.68,
      geo_region: "east asia"
    }]
  
    expect(wrapper.instance().filterCloudByRegion('east asia', mockClouds)).toEqual(expectedResult)
    expect(wrapper.instance().filterCloudByRegion('invalid region', mockClouds)).toEqual([])
  })
  
  test('should filter cloud list by cloud name and region and set filteredItems state', () => {
    wrapper = shallow(<CloudList />)
    wrapper.setState({
      clouds: mockClouds,
      selectedCloud: 'aws',
      selectedRegion: 'north america'
    })
  
    // searching with valid cloud name and region
    const expectedResult = [{
      cloud_description: "United States, California - Amazon Web Services: N. California",
      cloud_name: "aws-us-west-1",
      geo_latitude: 37.78,
      geo_longitude: -122.42,
      geo_region: "north america"
    }]
  
    wrapper.instance().filterCloudByNameAndOrRegion()
    expect(wrapper.state('filteredItems')).toEqual(expectedResult)
  
    // searching with only valid cloud name
    wrapper.setState({
      clouds: mockClouds,
      selectedCloud: 'aws',
      selectedRegion: ''
    })
    wrapper.instance().filterCloudByNameAndOrRegion()
    expect(wrapper.state('filteredItems')).toEqual(expectedResult)
  
    // searching with valid cloud name and invalid region
    wrapper.setState({
      clouds: mockClouds,
      selectedCloud: 'aws',
      selectedRegion: 'invalid region'
    })
    wrapper.instance().filterCloudByNameAndOrRegion()
    expect(wrapper.state('filteredItems')).toEqual([])
  
    // searching with invalid cloud name and region
    wrapper.setState({
      clouds: mockClouds,
      selectedCloud: 'noname',
      selectedRegion: 'invalid region'
    })
    wrapper.instance().filterCloudByNameAndOrRegion()
    expect(wrapper.state('filteredItems')).toEqual([])  
  })
  
  test('should get a correct full name based on the passed value', () => {
    wrapper = shallow(<CloudList />)
  
    expect(wrapper.instance().getCloudFullName('aws')).toEqual('Amazon Web Services')
    expect(wrapper.instance().getCloudFullName('azure')).toEqual('Microsoft Azure')
    expect(wrapper.instance().getCloudFullName('google')).toEqual('Google Cloud')
    expect(wrapper.instance().getCloudFullName('do')).toEqual('DigitalOcean')
    expect(wrapper.instance().getCloudFullName('upcloud')).toEqual('UpCloud')
    expect(wrapper.instance().getCloudFullName('unmatchedname')).toEqual('unmatchedname')
  })
})

describe('Populate the cloud options', () => {
  test('should set cloudNameOptions state value with correct formatted data when setCloudOptions method is called and cloud list has more than 0 item', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    wrapper.setState({ clouds: mockClouds2 })
    wrapper.instance().forceUpdate()

    const expectedFormat = [
      { value: 'google' , displayName: 'Google Cloud' },
      { value: 'azure' , displayName: 'Microsoft Azure' },
      { value: 'aws' , displayName: 'Amazon Web Services' }
    ]
    
    wrapper.instance().setCloudOptions()
    expect(wrapper.state('cloudNameOptions')).toEqual([emptyOption, ...expectedFormat])
  })

  test('should set cloudNameOptions state value with empty formatted data when setCloudOptions method is called and cloud list has 0 item', () => {
    wrapper = shallow(<CloudList />)
    wrapper.setState({ clouds: [] })
    wrapper.instance().forceUpdate()
    
    wrapper.instance().setCloudOptions()
    expect(wrapper.state('cloudNameOptions')).toEqual([emptyOption])
  })
})

describe('Populate the region options', () => {
  test('should set regionOptions state value with correct formatted data when setRegionOptions method is called and cloud list has more than 0 item', () => {
    wrapper = shallow(<CloudList />)
    wrapper.setState({ clouds: mockClouds2 })
    wrapper.instance().forceUpdate()

    const expectedFormat = [
      { value: mockClouds[0].geo_region, displayName: 'Europe' },
      { value: mockClouds[1].geo_region, displayName: 'East Asia' },
      { value: mockClouds[2].geo_region, displayName: 'North America' }  
    ]
    
    wrapper.instance().setRegionOptions()
    expect(wrapper.state('regionOptions')).toEqual([emptyOption, ...expectedFormat])
  })

  test('should set regionOptions state value with empty formatted data when setRegionOptions method is called and cloud list has 0 item', () => {
    wrapper = shallow(<CloudList />)
    wrapper.setState({ clouds: [] })
    wrapper.instance().forceUpdate()
    
    wrapper.instance().setRegionOptions()
    expect(wrapper.state('regionOptions')).toEqual([emptyOption])
  })
})

describe('handle change methods', () => {
  test('should update the selectedCloud state when handleCloudNameChange method is called', () => {
    wrapper = shallow(<CloudList />)
    wrapper.instance().handleCloudNameChange('aws')

    expect(wrapper.state('selectedCloud')).toEqual('aws')
  })

  test('should update the selectedRegion state when handleRegionChange method is called', () => {
    wrapper = shallow(<CloudList />)
    wrapper.instance().handleRegionChange('region')

    expect(wrapper.state('selectedRegion')).toEqual('region')
  })

  test('should update the selectedCity state correctly', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    const citySelect = wrapper.find('select#city')
    citySelect.simulate('change', { target: { value: 'city' } })

    expect(wrapper.state('selectedCity')).toEqual('city')
  })

  test('should call handleCityChange method when selecting a city option', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    wrapper.instance().handleCityChange = jest.fn()
    wrapper.instance().forceUpdate()
    const citySelect = wrapper.find('select#city')
    citySelect.simulate('change', { target: { value: 'city' } })

    expect(wrapper.instance().handleCityChange).toHaveBeenCalledTimes(1)
  })
})

describe('distance calculation', () => {
  test('should call getDistanceFromLatLonInKm helper function when calculateDistances method is called and filteredItems state has more than 1 item in array', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    const calDistanceSpy = jest.spyOn(helper, 'getDistanceFromLatLonInKm')
    wrapper.setState({ filteredItems: mockClouds })
    wrapper.instance().calculateDistances()

    expect(calDistanceSpy).toHaveBeenCalled()
    calDistanceSpy.mockRestore()
  })

  test('should not call getDistanceFromLatLonInKm helper function when calculateDistances method is called and filteredItems state has 1 item in array', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    const calDistanceSpy = jest.spyOn(helper, 'getDistanceFromLatLonInKm')
    wrapper.setState({ filteredItems: [mockClouds[0]] })
    wrapper.instance().calculateDistances()

    expect(calDistanceSpy).not.toHaveBeenCalled()
    calDistanceSpy.mockRestore()
  })

  test('should add distance propert to the returned cloud list if filteredItems state\'s length is more than 1 when calculateDistances method is called', async () => {
    wrapper = shallow(<CloudList />)
    await wrapper.instance().componentDidMount()
    wrapper.setState({ filteredItems: mockClouds })
    const result = wrapper.instance().calculateDistances()

    result.forEach((item: Cloud) =>{
      expect(item).toHaveProperty('distance')
    })
  })

  test('should sort the cloud list by distance in ascending order when calling sortDistanceAscen method with cloud list length is longer than 1', () => {
    wrapper = shallow(<CloudList />)
    const mockDistances = [
      {
        cloud_name: "google-europe-north1",
        distance: 100
      },
      {
        cloud_name: "azure-japaneast",
        distance: 1
      },
      {
        cloud_name: "aws-us-west-1",
        distance: 10
      }
    ]
    wrapper.instance().calculateDistances = jest.fn().mockReturnValue(mockDistances)
    wrapper.instance().forceUpdate()

    const expectedResult = [
      {
        cloud_name: "azure-japaneast",
        distance: 1
      },
      {
        cloud_name: "aws-us-west-1",
        distance: 10
      },
      {
        cloud_name: "google-europe-north1",
        distance: 100
      }
    ]

    wrapper.instance().sortDistanceAscen()
    expect(wrapper.state('filteredItems')).toEqual(expectedResult)
  })

  test('should just return the same cloud list when calling sortDistanceAscen method with cloud list length is 1', () => {
    wrapper = shallow(<CloudList />)
    const mockDistance = [{ cloud_name: "azure-japaneast", distance: 10000 }]
    wrapper.instance().calculateDistances = jest.fn().mockReturnValue(mockDistance)
    wrapper.instance().forceUpdate()

    wrapper.instance().sortDistanceAscen()
    expect(wrapper.state('filteredItems')).toEqual(mockDistance)
  })
})
