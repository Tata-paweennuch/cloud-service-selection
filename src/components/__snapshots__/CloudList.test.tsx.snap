// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when isLoading is true 1`] = `
<div
  className="content-wrapper"
>
  <p>
    Loading ....
  </p>
</div>
`;

exports[`should render correctly with cloud selection 1`] = `
<div
  className="content-wrapper"
>
  <div
    className="dropdown"
  >
    <label
      htmlFor="cloud"
    >
      Select cloud
    </label>
    <select
      id="cloud"
      name="cloud"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        -- select an option --
      </option>
      <option
        value="google"
      >
        Google Cloud
      </option>
      <option
        value="azure"
      >
        Microsoft Azure
      </option>
      <option
        value="aws"
      >
        Amazon Web Services
      </option>
    </select>
  </div>
  <div
    className="dropdown"
  >
    <label
      htmlFor="region"
    >
      Select region
    </label>
    <select
      id="region"
      name="region"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        -- select an option --
      </option>
      <option
        value="europe"
      >
        Europe
      </option>
      <option
        value="east asia"
      >
        East Asia
      </option>
      <option
        value="north america"
      >
        North America
      </option>
    </select>
  </div>
  <div
    className="dropdown"
  >
    <label
      htmlFor="city"
    >
      Select city
      <small>
        (Listed from nearest to farest)
      </small>
    </label>
    <select
      id="city"
      name="city"
      onChange={[Function]}
      value=""
    >
      <option
        value="Europe, Finland - Google Cloud: Finland"
      >
        Europe, Finland - Google Cloud: Finland
         * Nearest
      </option>
      <option
        value="Asia, Japan - Azure: Japan East"
      >
        Asia, Japan - Azure: Japan East
        
      </option>
      <option
        value="United States, California - Amazon Web Services: N. California"
      >
        United States, California - Amazon Web Services: N. California
        
      </option>
    </select>
  </div>
</div>
`;
